src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.DateTime.cs

```diff
@@ -0,0 +1,290 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    public ref partial struct Utf8JsonWriter
+    {
+        /// <summary>
+        /// Writes the property name and <see cref="DateTime"/> value (as a JSON string) as part of a name/value pair of a JSON object.
```

Comment:
I think it would be useful to say how the DateTime will be written as a string. I would guess it is ISO8061, but that would be useful intel in case someone wants a date written in a different format.

Reply:
We format based on the default `Utf8Formatter` behavior (in this case G). I will add that to the comments. 
https://github.com/dotnet/corefx/blob/069139a82c97ee045c7408db8483b0caaf062484/src/Common/src/CoreLib/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Date.cs#L140

We currently do not support `DateTimeKind` and if we were to extend the formatting functionality, it would very likely be based on `StandardFormat` which we would then route to `Utf8Formatter`

Reply:
That's a pretty bad format.  I'd pretty well recommend dropping the method over having that format.

"O" is the only sensible wire format in Utf8Formatter's list.

Reply:
No one wants `05/25/2017 10:30:15`. You should write ISO 8601. Also put careful thought into how you handle DateTime.Kind and DateTimeOffset.Offset.

Newtonsoft.Json outputs dates that are very similar to System.Xml. I recommend you follow. Everyone uses dates, they are very easy to get wrong and systems are very sensitive when you do get it wrong. You also can't go changing the result. I can tell you know that if you go it alone it will come back to haunt you.

Reply:
> Also put careful thought into how you handle DateTime.Kind and DateTimeOffset.Offset.

> "O" is the only sensible wire format in Utf8Formatter's list.

The StandardFormat "O" handles DateTime.Kind.

**We have three options here:**
1) Only provide a single `Write(DateTime)` overload that uses the "O" StandardFormat (instead of default). Now, `Utf8JsonWriter` deviates from `Utf8Formatter` by default.
2) Provide Write(DateTime) which uses the default StandardFormat and provide additional overloads that support other StandardFormats (like "O")
3) Provide a single Write(DateTime) overload that uses the "O" StandardFormat (instead of default), for now, but leave the option to provide other DateTime overloads in the future. Now, `Utf8JsonWriter` deviates from `Utf8Formatter` by default.

Based on the feedback, it looks like option 2 (what we have today) isn't the right solution.

Is it fine for `Utf8JsonWriter` to be opinionated and deviate from the default `Utf8Formatter`?

> You should write ISO 8601

Is the "O" format compatible to ISO 8601? Something like: `2017-06-12T05:30:45.7680000-07:00`? If so, I can change to that format by default.

Reply:
I recommend you produce a result for `DateTime` that is the same from [XmlConvert.ToString(DateTime, XmlDateTimeSerializationMode)](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlconvert.tostring?view=netframework-4.7.2#System_Xml_XmlConvert_ToString_System_DateTime_System_Xml_XmlDateTimeSerializationMode_) when `XmlDateTimeSerialization` is `RoundtripKind`.

Reply:
> Now read those JSON array values back to DateTimes (once with the raw reader, once with the JsonDocument). The DateTime.Now value should have a Kind of Unspecified, the DateTime.UtcNow should have a Kind of Utc.

The JsonReader/JsonDocument does not expose any GetDateTime API atm. However, if we add them, it makes sense to use the same standard format as the writer to ensure the round trip matches.

Reply:
> I recommend you produce a result for `DateTime` that is the same from [XmlConvert.ToString(DateTime, XmlDateTimeSerializationMode)](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlconvert.tostring?view=netframework-4.7.2#System_Xml_XmlConvert_ToString_System_DateTime_System_Xml_XmlDateTimeSerializationMode_) when `XmlDateTimeSerialization` is `RoundtripKind`.

With "O", it is the same as long as the seconds fraction is not 0. Otherwise, in XmlConvert, the .0000000 is omitted. The output matches what date.ToString("O") returns.

```C#
var date = DateTime.Now;
var dateUtc = DateTime.UtcNow;
byte[] destination = new byte[100];

Assert.True(Utf8Formatter.TryFormat(date, destination, out int bytesWritten, new Buffers.StandardFormat('O')));
string expectedStr = XmlConvert.ToString(date, XmlDateTimeSerializationMode.RoundtripKind);
string actualStr = Encoding.UTF8.GetString(destination.AsSpan(0, bytesWritten));
Assert.Equal(date.ToString("O"), actualStr);    // 2019-01-12T02:59:04.3282788-08:00
Assert.Equal(actualStr, expectedStr);

Assert.True(Utf8Formatter.TryFormat(dateUtc, destination, out bytesWritten, new Buffers.StandardFormat('O')));
expectedStr = XmlConvert.ToString(dateUtc, XmlDateTimeSerializationMode.RoundtripKind);
actualStr = Encoding.UTF8.GetString(destination.AsSpan(0, bytesWritten));
Assert.Equal(dateUtc.ToString("O"), actualStr);    // 2019-01-12T11:08:06.9461261Z
Assert.Equal(actualStr, expectedStr);

date = new DateTime(2019, 1, 1);
Assert.True(Utf8Formatter.TryFormat(date, destination, out bytesWritten, new Buffers.StandardFormat('O')));
expectedStr = XmlConvert.ToString(date, XmlDateTimeSerializationMode.RoundtripKind);
actualStr = Encoding.UTF8.GetString(destination.AsSpan(0, bytesWritten));
Assert.Equal(date.ToString("O"), actualStr);    // 2019-01-01T00:00:00.0000000
Assert.Equal(actualStr, expectedStr);   // Fails: 2019-01-01T00:00:00
```

Is this something we need to address (either fixing `Utf8Formatter` OR explicitly accounting for it in `Utf8JsonWriter`? I am assuming we can't change the behavior of `date.ToString()`, and it makes sense for `Utf8Formatter` to be consistent with it.

Reply:
@krwq, why does `XmlConvert.ToString` trim out the trailing 0s and differ from `date.ToString("O")`?

Reply:
It has been like that since the code was written in ~2004. Both `O` format and the custom XmlConvert formatting for date times was added around the same time. Looks like it was just a problem with people not talking to each other enough.

Datetimes are often rounded to the whole second, so omitting the fraction leads to more compact wire-format and it does not break round tripping. I guess it is why the Xml team choose to omit it. The wire-size was on top of their mind.

Reply:
> Datetimes are often rounded to the whole second, so omitting the fraction leads to more compact wire-format and it does not break round tripping. I guess it is why the Xml team choose to omit it. The wire-size was on top of their mind.

But it isn't always omitted. It is only removed if the fraction part is 0.

The question remains, what should JsonWriter do? Be consistent with Utf8Formatter and date.ToString() or explicitly omit .0000000 whenever the fraction part is zero?

Reply:
> But it isn't always omitted. It is only removed if the fraction part is 0.

Right, it is omitted only when it does not break round-tripping. It produces smallest valid string that round-trips by omitting insignificant trailing zeros, e.g. `XmlConvert.ToString(new DateTime(2019, 1, 1, 2, 42, 42, 100), XmlDateTimeSerializationMode.RoundtripKind)` produces `2019-01-01T02:42:42.1`. It is very similar to how floating point number formatting works - it omits insignificant trailing zeros as well.

> what should JsonWriter do?

I think it should be produce smallest valid string that round-trips.

Reply:
+1 smallest valid result.

> The JsonReader/JsonDocument does not expose any GetDateTime API atm

I think that is something that will need to improve but it is outside the scope of this PR.

Reply:
Filed an issue to address this: https://github.com/dotnet/corefx/issues/34576
Either I have a custom implementation of `Utf8Formatter.TryFormt` (that doesn't write the fractional part) or write to a scratch buffer using `Utf8Formatter` and exclude it when writing to an IBufferWriter. Otherwise, we would need to expose this functionality publicly as a new `StandardFormat` support.

Reply:
If what you're going to produce is the same as "O" but with some trailing zeros removed, I'd change it to "O" formatting before merge and then make it more optimal later.  Changing the format completely seems like a weird thing to defer.

Reply:
> I'd change it to "O" formatting before merge and then make it more optimal later. 

And... I see that's what you did, but the format is not near this line, so I didn't see it out on this view.